### ISSUE TITLE
Discrepancy between batch processing and single image inference results in Qwen3VLEmbedder

### ISSUE THREAD

**Issue by duybui1911** [QwenLM/Qwen3-VL-Embedding#31]
Hi Qwen team,

When running the `Qwen3VLEmbedder` model, there is a noticeable difference in the feature embeddings generated when images are processed in a batch compared to when they are processed individually (single loop).

```python
# model: 2B, float16
# fw: transformers==4.57.1

data = []
for img_path in paths:
    image = Image.open(img_path).convert("RGB")
    image = image.resize((384, 384))
    data.append({"image": image})

# batch process
batch_process = model.process(inputs=data)

# single process
single_process = []
for item in data:
    feature = model.process(inputs=[item])
    single_process.append(feature[0])

for i in range(len(batch_process)):
    diff = torch.abs(batch_process[i] - single_process[i])
    print(f"Image {i} diff: {torch.sum(diff)}")
```

The `torch.sum(diff.abs())` calculation shows non-zero values, indicating numerical divergence between the two methods.

```text
Image 0 diff: 0.2978515625
Image 1 diff: 0.224853515625
Image 2 diff: 0.29052734375
Image 3 diff: 0.1651611328125
Image 4 diff: 0.396484375
Image 5 diff: 0.252685546875
Image 6 diff: 0.8564453125
Image 7 diff: 0.305908203125
Image 8 diff: 0.46484375
Image 9 diff: 0.31591796875
```

Is it possible that there is information/position sharing (e.g., via attention mechanisms or padding) between images when they are processed in a batch, causing the embeddings to differ from single-image processing?

Thank you!

**Comment by 8igfive** [comment]
Thanks for raising this question. We have not yet specifically investigated the impact of this behavior in depth. However, for the issue itself, we believe it is likely related to nondeterminism during inference in Transformers models. For a detailed discussion of this phenomenon, you may find the following article helpful: [https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/](https://thinkingmachines.ai/blog/defeating-nondeterminism-in-llm-inference/)

**Comment by duybui1911** [comment]
Ah, I understand, thank you for your information.
The difference is very small in the 2048 dimensions, and I was taking the absolute value; I have switched the metric to cosine similarity and obtained similarities from 0.999 -- 1. It is fine and not an issue.
Thank you for your reply.
